src/webapi/html_elements/select.rs

```diff
@@ -0,0 +1,181 @@
+use webapi::element::{Element, IElement};
+use webapi::event_target::{EventTarget, IEventTarget};
+use webapi::html_element::{HtmlElement, IHtmlElement};
+use webapi::node::{INode, Node};
+use webcore::try_from::TryInto;
+use webcore::value::Reference;
+use webapi::html_collection::HtmlCollection;
+use webapi::html_elements::OptionElement;
+
+/// The HTML <select> element represents a control that provides a menu of options.
+///
+/// [(JavaScript docs)](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/select)
+// https://html.spec.whatwg.org/#htmlselectelement
+#[derive(Clone, Debug, PartialEq, Eq, ReferenceType)]
+#[reference(instance_of = "HTMLSelectElement")]
+#[reference(subclass_of(EventTarget, Node, Element, HtmlElement))]
+pub struct SelectElement(Reference);
+
+impl IEventTarget for SelectElement {}
+impl INode for SelectElement {}
+impl IElement for SelectElement {}
+impl IHtmlElement for SelectElement {}
+
+impl SelectElement {
+    /// Returns the `Some(index)` of the first selected item, if any, or `None` if there is no selected item.
+    // https://html.spec.whatwg.org/#the-select-element:dom-select-selectedindex
+    pub fn selected_index(&self) -> Option<i32> {
+        let si = js! (
+            return @{self}.selectedIndex;
+        ).try_into().unwrap();
+        if si < 0 {
+            None
+        } else {
+            Some(si)
+        }
+    }
+
+    /// Change selected index to the given value.
+    // https://html.spec.whatwg.org/#the-select-element:dom-select-selectedindex
+    pub fn set_selected_index(&self, selected_index: Option<i32>) {
+        let selected_index = selected_index.unwrap_or(-1);
+        js!{
+            @(no_return)
+            @{self}.selectedIndex = @{selected_index};
+        }
+    }
+
+    /// Returns the `Some(value)` of the first selected item, if any, or `None` if there is no selected item.
+    // https://html.spec.whatwg.org/#the-select-element:dom-select-value
+    pub fn value(&self) -> Option<String> {
+        js!(
+            var self = @{self};
+            if (self.selectedIndex < 0) {
+                return null;
+            } else {
+                return self.value;
+            }
+        ).try_into().unwrap()
+    }
+
+    /// Change the selected value to the given value.
+    // https://html.spec.whatwg.org/#the-select-element:dom-select-value
+    pub fn set_value(&self, value: Option<String>) {
```

Comment:
Hmm.... I'm not convinced we actually need to take the `Option<String>` here? I mean, there is already `set_selected_index(None)` that one can use. I know that `value` returns `Option<String>`, but that's normal for a getter; setters usually don't take an `Option`.

Reply:
I think it should be consistent with the getter, and I also think it's very useful to be able to (easily and intuitively) unset the value.

Reply:
Having the setter take `Option<String>` signals that it is possible for it to *not* have a value at all.

If it takes `String` then that implies that it always has a value, which is not true. And then if the user *wants* to unset the value, they need to use a completely different API, which doesn't sound good to me.

Reply:
Also, the decision for `get_value` to return an `Option<String>` (rather than requiring the user to use `get_selected_index`) was to prevent mistakes, and make the API more intuitive. I believe the same argument applies to `set_value`

Reply:
Okay, fine, you can keep it `Option<String>`; just:

1. Call `set_selected_index` inside of setting it through `js!`.
2. In the `Some` case add a `debug_assert!` after setting the value to verify that `self.selected_index().is_some()` - if we're going to make the interface more bullet proof let's go all the way (otherwise both `None` and `Some("insert gibberish here")` will result in the value being unset). Or perhaps we could instead always check that and return a `Result`? (Basically make setting an invalid value into an error.) I don't like messing with those APIs too much, but they *are* pretty poorly designed. (@Pauan Your thoughts?)

Reply:
Hmmm... these are my thoughts on it:

1. Add in `raw_value` and `set_raw_value` which only work with `String` (not `Option<String>`), and they never error. So they're just a thin wrapper over the underlying DOM APIs.

2. Change `get_selected_index` and `set_selected_index` to use `u32` (with a `debug_assert` to ensure that it doesn't wrap around to negative)

3. Change `value` and `set_value` to call `get_selected_index` + `raw_value`, and `set_selected_index` + `set_raw_value`

4. Change `set_value` to return `Result`

Reply:
Hmm... sounds good to me!

@limira Let's do what Pauan proposed. For the error type returned by `value()` you can create a dedicated type for it and return it when the user passes an unknown value. (`UnknownVariantError`, perhaps?)

Reply:
I think `UnknownValueError` or `InvalidValueError` is more intuitive, since you're setting a value which doesn't exist.

Reply:
Those also sound fine to me!